### PR TITLE
Massive Overhaul (By CarJem and campbellsonic)

### PR DESCRIPTION
**Pull request by CarJem**
There might be minimal leftovers in the code but it is mostly clean. Notably, this pull will be able to close Issue #5. 

~~EDIT: There is a personal tweak I made where the FG Low and High do not hide when a layer besides the main three are selected. I can either revert this or add yet another toggle.~~

EDIT 2: Added that in the latest commit

Also, I never said where that "Copy Between Scenes" Toggle is. It's in the about menu.

### CODE REVIEW

**Review by OtherworldBob — COMMENTED** [review]
I think this variable is unused, and should be removed to avoid confusion.

File: ManiacEditor/Editor.cs [hunk 1]
```diff
@@ -62,7 +62,11 @@ public partial class Editor : Form, IDrawArea
         private EntitiesToolbar entitiesToolbar = null;
 
         internal Dictionary<Point, ushort> TilesClipboard;
-        private List<EditorEntity> entitiesClipboard;
+        private List<EditorEntity> entitiesClipboard;
+        internal int SelectedTilesCount;
+        internal int SelectedTileX;
+        internal int SelectedTileY;
+        internal bool forceCopyUnlock;
```
Comment (OtherworldBob) [inline ManiacEditor/Editor.cs]: I think this variable is unused, and should be removed to avoid confusion.

**Review by OtherworldBob — COMMENTED** [review]
Just for consistency, could you capitalise the first letter here, and in the other new option?

File: ManiacEditor/Properties/Settings.settings [hunk 2]
```diff
@@ -1,18 +1,24 @@
...
+    <Setting Name="DataDirectories" Type="System.Collections.Specialized.StringCollection" Scope="User">
+      <Value Profile="(Default)" />
+    </Setting>
+    <Setting Name="UpgradeRequired" Type="System.Boolean" Scope="User">
+      <Value Profile="(Default)">True</Value>
+    </Setting>
+    <Setting Name="GamePath" Type="System.String" Scope="User">
+      <Value Profile="(Default)" />
+    </Setting>
+    <Setting Name="forceCopyUnlock" Roaming="true" Type="System.Boolean" Scope="User">
```
Comment (OtherworldBob) [inline ManiacEditor/Properties/Settings.settings]: Just for consistency, could you capitalise the first letter here, and in the other new option?

Comment (OtherworldBob) [inline ManiacEditor/Properties/Settings.settings]: Just for consistency, could you capitalise the first letter here, and in the other new option?

**Review by OtherworldBob — COMMENTED** [review]
Would be useful to give more descriptive names to the labels you are now using, and simply deleting the placeholders.

File: ManiacEditor/Editor.designer.cs [hunk 3]
```diff
@@ -931,7 +969,11 @@ private void InitializeComponent()
         private System.Windows.Forms.ToolStripStatusLabel _baseDataDirectoryLabel;
         private System.Windows.Forms.ToolStripMenuItem openDataDirectoryToolStripMenuItem;
         private System.Windows.Forms.ToolStripSeparator toolStripSeparator10;
-        private System.Windows.Forms.ToolStripButton ReloadToolStripButton;
+        private System.Windows.Forms.ToolStripButton ReloadToolStripButton;
+        private System.Windows.Forms.ToolStripStatusLabel toolStripStatusLabel2;
```
Comment (OtherworldBob) [inline ManiacEditor/Editor.designer.cs]: Would be useful to give more descriptive names to the labels you are now using, and simply deleting the placeholders.

**Review by OtherworldBob — COMMENTED** [review]
Hey nice work, thanks for sending over a PR!
It would be great if you could tidy up the unused events/click handlers that have snuck in. But that's no big issue.

Thanks for pointing out the extra options, I don't think I'd have found them in the About box for a while. Please feel free to add a full-on new Options dialog and menu option, if you feel that'd make more sense.

Three little issues I noticed with the Tile Count display:
* If you have selected tiles, and you switch layers, the count only resets once the mouse re-enters the editor area. Not a huge deal I think, just a little weird.
* If you select some tiles and hit delete, the count only resets the next time you click.
* If you use the Ctrl+Click mechanism to remove tiles from a multiple selection, the count incorrectly resets to 0. Adding further tiles seems to work fine though.

Don't forget to give yourself a credit in the About box too!

**Comment by CarJem** [comment]
@OtherworldBob Yeah, I haven't exactly worked out the flaws with it yet, still trying to figure out where to make the calls. Not to mention this is my first time with C# not to mention my first pull request so I will make these tweaks.

Also, campbellsonic has created a Clipboard that works between instances, that will be hopefully merged into this merge under your acceptance. Also, can I add the icons with this merge? They're long in the coming.

**Comment by CarJem** [comment]
OK, I fixed all the problems mentioned and uploaded the changes, hopefully, soon I will upload the Clipboard between Instances of Maniac. Then this pull can be finalized!

**Comment by OtherworldBob** [comment]
Thanks for making those updates.
I think the cross-instance Clipboard changes should be made separately, already plenty going on here!

For now, I'd probably just suggest removing the currently unused imagery/icons; they can be reinstated later if needs be. 

Then yep, let's hit merge and get this stuff in.

**Comment by OtherworldBob** [comment]
You keep adding stuff! (This is a good thing, don't stop!)
I've removed some of the unused stuff, just to keep things a bit tidier more than anything. 

Unfortunately I broke the Tile tracker again:
incorrectposition
If there are largish gaps between the tiles selected, the label can sometimes point to a space in between. Though weirdly, this doesn't happen every time.

I reckon the simplest fix would be to display the position of the top-left most selected tile.
Later we (or indeed your good self, it can wait for a later go around) could add a selected size indicator, pushing all the way to the bottom-right most tile. 
Still a bit unusual for groups of tiles that are split over a large area, but the best compromise so far I think.

**Comment by CarJem** [comment]
@OtherworldBob Think about the Pixel Selection this way, without it, it would have been hours for me to calculate for very specific entities like water, which calculates in 16's instead of tiles, on the other hand, some entities like breakablewall and invisiblewall work like that. That's why I implemented it, in fact, I was able to use it practically the other day when working on my Lost Hidden Palace Zone Mod, I was able to set up near perfect water boundaries with this small detail in a lot less time than it would of to count by hand. Which is why I am sad about its removal. I request that it may be reimplemented.

Also, the tile tracker is still very crude considering I have no idea where all the calls are being made when selections are being done. Will continue to investigate this, but as it stands it is a useful feature for now.

Also, the location of the tile tracker is something to largely be worked on, as it is right now, it is really hard to read from that will be adjusted.

**Comment by OtherworldBob** [comment]
I'm lacking time at the moment, and the other branch is now so far ahead in terms of features that this PR is just massively huge. If I ever have the time to help contribute to this project again, I will!
Closing for now.

**Pull request closed without merging**